riscv/runtime/src/allocator.rs

```diff
@@ -0,0 +1,65 @@
+//! A very simple global allocator.
+//!
+//! Allocates on a global array and never deallocates.
+
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    cell::Cell,
+    ptr::{self, addr_of},
+};
+
+struct FixedMemoryAllocator<const SIZE: usize> {
+    mem_buffer: [u8; SIZE],
+    next_available: Cell<usize>,
+}
+
+impl<const SIZE: usize> FixedMemoryAllocator<SIZE> {
+    const fn new() -> Self {
+        Self {
+            mem_buffer: [0; SIZE],
+            next_available: Cell::new(0),
+        }
+    }
+}
+
+unsafe impl<const SIZE: usize> GlobalAlloc for FixedMemoryAllocator<SIZE> {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        // Start address of the allocation array:
+        let array_start = addr_of!(self.mem_buffer) as usize;
+
+        // Address of the next free space:
+        let next_ptr = array_start + self.next_available.get();
+
+        // Align the pointer.
+        let aligned_ptr = (next_ptr + layout.align() - 1) & !(layout.align() - 1);
+
+        // Where this allocated space ends:
+        let end_of_allocation_ptr = aligned_ptr + layout.size();
+
+        // Calculates where the next allocation with start:
+        let new_next_available = end_of_allocation_ptr - array_start;
+
+        if new_next_available <= SIZE {
+            self.next_available.set(new_next_available);
+            aligned_ptr as *mut u8
+        } else {
+            ptr::null_mut()
```

Comment:
What happens if this returns null? We should not be able to generate a valid proof in case of out-of-memory, I think.

Reply:
Depends on the high level API used. If it can fail, returns `Result::Err(...)` or `Option::None`, but if it can't fail, like in `Vec` or other containers, then the function marked with `#[alloc_error_handler]` is called, which panics.

Reply:
So, yeah, it will fail unless the program is coded to handle out of memory and still succeed.